pkg/monitor/loki_upload.go

```diff
@@ -0,0 +1,285 @@
+// Includes a rudimentary retry mechanism.
+func pushBatch(client *http.Client, bearerToken string, lokiData LokiData) error {
+	batchCtr++
+	url := "https://logging-loki-openshift-operators-redhat.apps.cr.j7t7.p1.openshiftapps.com/api/logs/v1/openshift-trt/loki/api/v1/push"
+	headers := map[string]string{
+		"Content-Type":  "application/json",
+		"Authorization": "Bearer " + bearerToken,
+	}
+	jsonData, err := json.Marshal(lokiData)
+	if err != nil {
+
+		logrus.WithError(err).Error("error marshalling loki data")
+		return err
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+	if err != nil {
+		logrus.WithError(err).Error("error creating HTTP request")
+		return err
+	}
+
+	for key, value := range headers {
+		req.Header.Set(key, value)
+	}
+
+	return retry.OnError(
+		wait.Backoff{
+			Duration: 2 * time.Second,
+			Steps:    5,
+			Factor:   5.0,
+			Jitter:   0.1,
+		},
+		func(err error) bool {
+			// re-try on any error
+			logrus.WithError(err).Warningf("error occurred on batch %d, re-trying", batchCtr)
+			return true
+		},
+		func() error {
+			logrus.Infof("pushing batch %d of %d intervals", batchCtr,
+				len(lokiData.Streams[0].Values))
+			response, err := client.Do(req)
+			if err != nil {
+				logrus.WithError(err).Error("error making HTTP request")
+				return err
+			}
+			defer response.Body.Close()
+
+			body, err := ioutil.ReadAll(response.Body)
+			if err != nil {
+				logrus.WithError(err).Error("error reading response body")
+				return err
+			}
+
+			status := response.StatusCode
+			if len(body) > 0 {
+				logrus.Infof("request body: %s", body)
+			}
+			if status >= 200 && status < 300 {
+				logrus.WithField("status", status).
+					Infof("%d log entries pushed to loki", len(lokiData.Streams[0].Values))
+			} else {
+				logrus.WithField("status", status).
+					Errorf("%d log entries failed to push to loki", len(lokiData.Streams[0].Values))
+				return fmt.Errorf("logs push to loki failed with http status %d", status)
+			}
+
+			return nil
+		})
+}
+
+// UploadIntervalsToLoki attempts to push the set of intervals from this portion of the job run up to Loki,
+// allowing for more efficient querying and searching in TRT tooling.
+func UploadIntervalsToLoki(intervals monitorapi.Intervals) error {
+
+	// Ensure we have appropriate env vars defined, if not exit with an error.
+	clientID, defined := os.LookupEnv("LOKI_SSO_CLIENT_ID")
+	if !defined {
+		return fmt.Errorf("LOKI_SSO_CLIENT_ID env var is not defined")
+	}
+	clientSecret, defined := os.LookupEnv("LOKI_SSO_CLIENT_SECRET")
+	if !defined {
+		return fmt.Errorf("LOKI_SSO_CLIENT_SECRET env var is not defined")
+	}
+	jobName, defined := os.LookupEnv("JOB_NAME")
+	if !defined {
+		return fmt.Errorf("JOB_NAME env var is not defined")
+	}
+	buildID, defined := os.LookupEnv("BUILD_ID")
+	if !defined {
+		return fmt.Errorf("BUILD_ID env var is not defined")
+	}
+
+	bearerToken, err := obtainSSOBearerToken(clientID, clientSecret, intervals)
```

Comment:
Any concerns about calling this for each batch?  Presume we won't have 10s or 100s of batches but if we did would it be a problem / unnecessary to call SSO api in a tight(ish) loop?